### DISCUSSION TITLE
Custom Cards maintanance / installation / updates

### DISCUSSION THREAD

**Discussion by dougmaitelli** [UI-Lovelace-Minimalist/UI#819] [Ideas]
Hey,

I am wondering, how is everyone handling custom cards updates? From what I see the only way is to manually update all the custom cards you have by copying the files from the repo every time there is a new change?
Any better way to handle this?

**Comment by stokkie90** [comment]
Currently there isn't. I also was thinking about this. Some options i'm thinking are:

- option in the configure menu to copy over ALL custom cards 
- option in the configure menu to select certain carts to copy over to the correct directory. 


@basbruss What do you think?

**Reply by basbruss** [reply]
> * option in the configure menu to select certain carts to copy over to the correct directory.

This seems a good option, as it will put less load on lower power systems as the first option. 

However, you can't as a user add custom alterations to the custom cards anymore, as they will be overridden.
Or they have to put them in separate files, which is basically the same as the system we have now in place.

So then we have to decide what custom card features we prefer most.

- Easy installing and updating through the integration.
- Endless customizability without losing local changes.

Edit: Normally custom_cards do not receive a lot of updates. It's now because we changed the whole framework that a lot of custom cards relied on.

**Reply by basbruss** [reply]
Btw maybe I have a solution to prevent overrides when a file is altered locally.

@stokkie90 Is it possible to build a workflow in python that scans local YAML files? If so we can let users add a tag to their files when they have altered them. So when the tag is found by the workflow, the integration does not need to copy over that certain card file. And when no tag is found it can safely update that custom card.

**Reply by stokkie90** [reply]
I think that should be possible. to do this correctly it might even be better to create like a default tag that we set when configuring a template. if that tag is not set, it's not managed by ulm.

For example it could be done with:

```yaml
#ulm_managed
---

custom_card_camera:
....
```

if that tag exists, the integration is allowed to overwrite the file in case of an update.

To do this we need to build in some kind of git clone mechanism or keep like a static list of custom cards. Downside of a static list is. you need to keep it up-to-date and depends on releases. up-to-date can be done during the release step i think. 

Notes for myself:
- figure out a way to get a list of all custom cards 
  - github api list dir 
  - static list as fallback 
- loop over enabled cards
  - check if dir exists
    - check if managed tag is set in the file
    - do md5sum check to see if there is a diff (optional)
    - overwrite
- do all these actions on start & reload & changing settings.

**Reply by basbruss** [reply]
That has some potential 😄 
https://pynative.com/python-yaml/#h-python-yaml-load-read-yaml-file this can allows us to make a dict of all custom cards in a for loop. Then we only need to add a string finder function to the loop.



> * do md5sum check to see if there is a diff (optional)

Will this be effective? If a custom card has received an update, the original (updated file) will always differ from the local one.

**Reply by stokkie90** [reply]
yeah not sure yet about the md5sum check. just overwriting it might be simpler :sweat_smile: 

about the tag we could also use:

```yaml
---
ulm_managed: true 

custom_card_camera:
..
```

that makes it easer to load the yaml. else we could use:

```python
with open("sample.yaml") as f:
    firstline = f.readline().rstrip()
    if "#ulm_managed" in firstline:
      """do magic here"""
```

**Comment by dougmaitelli** [comment]
Can't the user custom-cards and the packaged custom cards be kept in separate folders?
Then loading can take priority of the user ones.

**Comment by Neekster** [comment]
I would love automatic updates of custom cards, but I still want the ability to make my own custom cards.

I would also like to suggest a rename to "community cards" (or something similar) and reserve "custom cards" for ones made by the individual.

An auto update system could then download the community cards to a community cards folder, and custom cards can be manually added to the custom cards folder. This also allows for alterations of community cards by adding them to the custom cards folder so that your edits don't get overridden.

**Comment by stokkie90** [comment]
In case you are all interested: https://github.com/UI-Lovelace-Minimalist/UI/pull/888

i have a working version to do all of this in this PR. Would like to have you all to take a look at it.
